File: python/basics/math/50_power.py
'''
https://leetcode.com/problems/powx-n/

x**n = (x**2)**(n/2)

x**n = x. x**(n-1) = x. (x**2)**((n-1)/2)

#interesting if you want O(logN)
'''

def power(x:int,n:int):
    if n==0:
        return 1
    if n < 0:
        return 1/power(x, -n)       #mistake need to take care of -ve power
    
    if n%2==0:
        return power((x*x), n/2)
    else:
        return x * power((x*x), (n-1)/2)



x = 2.00000
n = 10
#Output: 1024.00000
result = power(x, n)
print("result: " + str(result))

x = 2.10000
n = 3
#output: 9.26100
result = power(x, n)
print("result: " + str(result))



x = 2.00000
n = -2
#Output: 0.25000
result = power(x, n)
print("result: " + str(result))

File: problem_solving/math/cpp/50_power.cpp
/*
https://leetcode.com/problems/powx-n/


*/

#include<iostream>

double my_pow(double x, int n) {
    if(n == 0 ) {
        return 1;
    }
    long long ln = n;   //#mistake need this to avoid issues with large -ve numbers
    if(ln < 0) {
        x = 1/x;
        ln = -ln;
    }
    if(ln % 2 == 0) {
        return my_pow(x*x, ln/2);
    } else {
        return x * my_pow(x, ln-1);
    }
}


double my_pow_iterative(double x, int n) {
    if(n == 0 ) {
        return 1;
    }
    long long ln = n;   //#mistake need this to avoid issues with large -ve numbers
    if(ln < 0) {
        x = 1/x;
        ln = -ln;
    }
    double result = 1.0;
    double current = x;
    while(ln > 0) {
        if(ln % 2 == 1) {
            result *= current;
        }
        current *= current;
        ln /= 2;   
    }
    return result;
}


int main() {
    double x = 2;
    int n = 10;
    auto result = my_pow(x, n);
    std::cout << "result: " << result << std::endl;
}